File: go/paycore/paymentchannels/paymentchannels.go
// Package paymentchannels is the thin, hand-written on-chain glue over the
// codama-generated payment-channels client in
// protocols/programs/paymentchannels. It provides PDA derivation, associated
// token derivation, voucher preimage bytes, and convenience instruction
// builders for the push-mode session flow (open + top_up).
//
// The instruction bytes and PDA derivations built here must stay
// byte-identical across the language SDKs so the on-chain program accepts
// them. In particular the production program id pinned here (GuoKrza...)
// overrides the IDL placeholder baked into the generated package, which is
// not deployed.
package paymentchannels

import (
	"bytes"
	"encoding/binary"
	"fmt"

	ag_binary "github.com/gagliardetto/binary"
	solana "github.com/gagliardetto/solana-go"

	generated "github.com/solana-foundation/pay-kit/go/protocols/programs/paymentchannels"
)

// ProgramID is the canonical payment-channels program id deployed to the
// network. The codama-generated package defaults its ProgramID var to the IDL
// placeholder "CQAyft83tN1w2bRofB5PZ79eVDU2xZUVo43LU1qL4zRg", which is NOT the
// production deployment; every PDA derivation and instruction built here uses
// this value instead.
const ProgramID = "GuoKrzaBiZnW5DvJ3yZVE7xHqbcBvaX9SH6P6Cn9gNvc"

// channelSeed is the channel PDA seed prefix.
const channelSeed = "channel"

// eventAuthoritySeed is the event-authority PDA seed prefix.
const eventAuthoritySeed = "event_authority"

// programPubkey is the parsed production program id used for derivation and
// instruction emission.
var programPubkey = solana.MustPublicKeyFromBase58(ProgramID)

func init() {
	// Pin the generated package's ProgramID to the production deployment so
	// any path that reads generated.ProgramID (e.g. Instruction.ProgramID())
	// observes GuoKrza... rather than the IDL placeholder.
	generated.SetProgramID(programPubkey)
}

// ProgramPubkey returns the parsed production program id.
func ProgramPubkey() solana.PublicKey {
	return programPubkey
}

// SetProgramID overrides the program id used for PDA derivation and instruction
// emission, for SDK consumers targeting a non-mainnet deployment (a devnet or
// localnet program is deployed at a different address). It also pins the
// generated package so Instruction.ProgramID() agrees. The default is the
// canonical mainnet ProgramID; callers on mainnet never need this.
func SetProgramID(id solana.PublicKey) {
	programPubkey = id
	generated.SetProgramID(id)
}

// Distribution is a single payout recipient and its basis-point share.
type Distribution struct {
	// Recipient is the wallet whose associated token account receives this
	// share when settled channel funds are distributed.
	Recipient solana.PublicKey
	// Bps is the recipient's share of distributed funds in basis points
	// (100 bps = 1%).
	Bps uint16
}

// OpenChannelParams carries the inputs required to build an Open instruction.
type OpenChannelParams struct {
	// Payer is the wallet funding the channel deposit; it signs the Open
	// transaction and is a channel PDA seed.
	Payer solana.PublicKey
	// Payee is the counterparty the channel pays out to; a channel PDA seed.
	Payee solana.PublicKey
	// Mint is the SPL token mint the channel escrows (e.g. USDC); a channel
	// PDA seed.
	Mint solana.PublicKey
	// AuthorizedSigner is the key allowed to sign vouchers against this
	// channel; a channel PDA seed.
	AuthorizedSigner solana.PublicKey
	// Salt distinguishes multiple channels sharing the same
	// payer/payee/mint/signer; encoded little-endian as the final channel
	// PDA seed.
	Salt uint64
	// Deposit is the initial escrow amount in token base units
	// (10^-6 USDC per unit for a 6-decimal mint).
	Deposit uint64
	// GracePeriod is the channel close grace period in seconds; the on-chain
	// program rejects zero.
	GracePeriod uint32
	// Recipients is the basis-point payout split applied when settled funds
	// are distributed.
	Recipients []Distribution

	// TokenProgram is the program owning Mint (SPL Token or Token-2022),
	// used to derive the payer and channel associated token accounts.
	TokenProgram solana.PublicKey

	// ProgramID is the payment-channels program targeted by this open. The
	// zero value resolves to the package program id (ProgramPubkey, or the
	// last SetProgramID override).
	ProgramID solana.PublicKey
}

// TopUpParams carries the inputs required to build a TopUp instruction.
type TopUpParams struct {
	// Payer is the wallet whose token account funds the top-up; it signs
	// the TopUp transaction.
	Payer solana.PublicKey
	// Channel is the channel PDA whose escrow receives the deposit.
	Channel solana.PublicKey
	// Mint is the SPL token mint the channel escrows.
	Mint solana.PublicKey
	// Amount is the additional deposit in token base units
	// (10^-6 USDC per unit for a 6-decimal mint).
	Amount uint64
	// TokenProgram is the program owning Mint (SPL Token or Token-2022),
	// used to derive the payer and channel associated token accounts.
	TokenProgram solana.PublicKey

	// ProgramID is the payment-channels program targeted by this top-up. The
	// zero value resolves to the package program id (ProgramPubkey, or the
	// last SetProgramID override).
	ProgramID solana.PublicKey
}

// resolveProgram resolves an optional per-call program id to the package
// program id when unset.
func resolveProgram(programID solana.PublicKey) solana.PublicKey {
	if programID.IsZero() {
		return programPubkey
	}
	return programID
}

// VoucherMessageBytes returns the 48-byte voucher preimage signed by the
// authorized signer: channelId (32) || cumulativeAmount as little-endian u64
// (offset 32) || expiresAt as little-endian i64 (offset 40). This is the exact
// Borsh layout of VoucherArgs.
func VoucherMessageBytes(channelID solana.PublicKey, cumulative uint64, expiresAt int64) ([]byte, error) {
	id := channelID.Bytes()
	if len(id) != 32 {
		return nil, fmt.Errorf("channel id must be exactly 32 bytes, got %d", len(id))
	}
	out := make([]byte, 48)
	copy(out[:32], id)
	binary.LittleEndian.PutUint64(out[32:40], cumulative)
	binary.LittleEndian.PutUint64(out[40:48], uint64(expiresAt))
	return out, nil
}

// FindChannelPDA derives the channel PDA from
// ["channel", payer, payee, mint, authorizedSigner, salt as little-endian u64]
// against the production program id.
func FindChannelPDA(payer, payee, mint, authorizedSigner solana.PublicKey, salt uint64) (solana.PublicKey, uint8, error) {
	return FindChannelPDAForProgram(payer, payee, mint, authorizedSigner, salt, programPubkey)
}

// FindChannelPDAForProgram derives the channel PDA against an explicit program
// id, for callers honoring a per-challenge programId.
func FindChannelPDAForProgram(payer, payee, mint, authorizedSigner solana.PublicKey, salt uint64, programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	saltLE := make([]byte, 8)
	binary.LittleEndian.PutUint64(saltLE, salt)
	addr, bump, err := solana.FindProgramAddress(
		[][]byte{
			[]byte(channelSeed),
			payer.Bytes(),
			payee.Bytes(),
			mint.Bytes(),
			authorizedSigner.Bytes(),
			saltLE,
		},
		resolveProgram(programID),
	)
	if err != nil {
		return solana.PublicKey{}, 0, fmt.Errorf("derive channel pda: %w", err)
	}
	return addr, bump, nil
}

// FindEventAuthorityPDA derives the event-authority PDA from
// ["event_authority"] against the production program id.
func FindEventAuthorityPDA() (solana.PublicKey, uint8, error) {
	return FindEventAuthorityPDAForProgram(programPubkey)
}

// FindEventAuthorityPDAForProgram derives the event-authority PDA against an
// explicit program id, for callers honoring a per-challenge programId.
func FindEventAuthorityPDAForProgram(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	addr, bump, err := solana.FindProgramAddress(
		[][]byte{[]byte(eventAuthoritySeed)},
		resolveProgram(programID),
	)
	if err != nil {
		return solana.PublicKey{}, 0, fmt.Errorf("derive event-authority pda: %w", err)
	}
	return addr, bump, nil
}

// BuildOpenInstruction derives the channel PDA, payer/channel ATAs, and
// event-authority PDA, then builds the Open instruction with every account set
// in the exact order the on-chain program expects, using the production
// program id.
func BuildOpenInstruction(params OpenChannelParams) (solana.Instruction, error) {
	programID := resolveProgram(params.ProgramID)
	channel, _, err := FindChannelPDAForProgram(params.Payer, params.Payee, params.Mint, params.AuthorizedSigner, params.Salt, programID)
	if err != nil {
		return nil, err
	}
	payerToken, _, err := solana.FindAssociatedTokenAddressWithProgram(params.Payer, params.Mint, params.TokenProgram)
	if err != nil {
		return nil, fmt.Errorf("derive payer token account: %w", err)
	}
	channelToken, _, err := solana.FindAssociatedTokenAddressWithProgram(channel, params.Mint, params.TokenProgram)
	if err != nil {
		return nil, fmt.Errorf("derive channel token account: %w", err)
	}
	eventAuthority, _, err := FindEventAuthorityPDAForProgram(programID)
	if err != nil {
		return nil, err
	}

	recipients := make([]generated.DistributionEntry, 0, len(params.Recipients))
	for _, entry := range params.Recipients {
		recipients = append(recipients, generated.DistributionEntry{
			Recipient: entry.Recipient,
			Bps:       entry.Bps,
		})
	}

	builder := generated.NewOpenInstructionBuilder().
		SetPayerAccount(params.Payer).
		SetPayeeAccount(params.Payee).
		SetMintAccount(params.Mint).
		SetAuthorizedSignerAccount(params.AuthorizedSigner).
		SetChannelAccount(channel).
		SetPayerTokenAccountAccount(payerToken).
		SetChannelTokenAccountAccount(channelToken).
		SetTokenProgramAccount(params.TokenProgram).
		SetSystemProgramAccount(solana.SystemProgramID).
		SetRentAccount(solana.SysVarRentPubkey).
		SetAssociatedTokenProgramAccount(solana.SPLAssociatedTokenAccountProgramID).
		SetEventAuthorityAccount(eventAuthority).
		SetSelfProgramAccount(programID).
		SetOpenArgs(generated.OpenArgs{
			Salt:        params.Salt,
			Deposit:     params.Deposit,
			GracePeriod: params.GracePeriod,
			Recipients:  recipients,
		})

	if _, err := builder.ValidateAndBuild(); err != nil {
		return nil, fmt.Errorf("build open instruction: %w", err)
	}
	return materialize(builder, builder.GetAccounts(), programID)
}

// BuildTopUpInstruction derives the payer/channel ATAs and builds the TopUp
// instruction with every account set in the exact order the on-chain program
// expects, using the production program id.
func BuildTopUpInstruction(params TopUpParams) (solana.Instruction, error) {
	payerToken, _, err := solana.FindAssociatedTokenAddressWithProgram(params.Payer, params.Mint, params.TokenProgram)
	if err != nil {
		return nil, fmt.Errorf("derive payer token account: %w", err)
	}
	channelToken, _, err := solana.FindAssociatedTokenAddressWithProgram(params.Channel, params.Mint, params.TokenProgram)
	if err != nil {
		return nil, fmt.Errorf("derive channel token account: %w", err)
	}

	builder := generated.NewTopUpInstructionBuilder().
		SetPayerAccount(params.Payer).
		SetChannelAccount(params.Channel).
		SetPayerTokenAccountAccount(payerToken).
		SetChannelTokenAccountAccount(channelToken).
		SetMintAccount(params.Mint).
		SetTokenProgramAccount(params.TokenProgram).
		SetTopUpArgs(generated.TopUpArgs{Amount: params.Amount})

	if _, err := builder.ValidateAndBuild(); err != nil {
		return nil, fmt.Errorf("build top_up instruction: %w", err)
	}
	return materialize(builder, builder.GetAccounts(), resolveProgram(params.ProgramID))
}

// materialize borsh-encodes a validated generated instruction implementation
// and returns a solana.GenericInstruction pinned to the production program id.
//
// Two generated-package quirks are handled here:
//   - The instruction implementation (*Open/*TopUp) is encoded directly so its
//     MarshalWithEncoder writes the program's real one-byte discriminator
//     (Open=1, TopUp=3). Wrapping it in the generated Instruction.Data() would
//     prepend a spurious NoTypeID-default byte, corrupting the on-chain data.
//   - The implementation is stored by value inside the generated Instruction,
//     so its Accounts() accessor type-asserts to a pointer-receiver interface
//     and panics; passing the builder's own GetAccounts() avoids that path.
//
// The result's ProgramID() is the resolved per-call program id (the production
// ProgramID by default, a SetProgramID override, or an explicit per-call
// ProgramID for a non-mainnet cluster).
func materialize(impl ag_binary.EncoderDecoder, accounts []*solana.AccountMeta, programID solana.PublicKey) (solana.Instruction, error) {
	buf := new(bytes.Buffer)
	if err := ag_binary.NewBorshEncoder(buf).Encode(impl); err != nil {
		return nil, fmt.Errorf("encode instruction data: %w", err)
	}
	return solana.NewInstruction(programID, accounts, buf.Bytes()), nil
}

File: python/src/pay_kit/protocols/mpp/_paymentchannels.py
"""On-chain glue for the payment-channels program.

Provides PDA derivation, associated token derivation, voucher preimage bytes,
and convenience instruction builders for the push-mode session flow
(``open`` + ``topUp``).

Instruction data and account metas are produced by the codama-py generated
client under :mod:`pay_kit.protocols.programs.paymentchannels` (rendered from
``idl/payment-channels.json`` by ``skills/pay-sdk-implementation/codegen``).
This module only adds what the IDL cannot express:

- The production program id (``GuoKrza...``) overrides the IDL placeholder
  (``CQAyft83tN1w2bRofB5PZ79eVDU2xZUVo43LU1qL4zRg``), which is not the deployed
  program; every PDA derivation and instruction built here uses ``GuoKrza...``.
  The generated PDA helpers pin the placeholder and take no program id
  parameter, so the event-authority derivation stays here.
- The channel PDA is not declared in the IDL's ``pdas`` section, so its
  derivation is hand-written here.

The payment-channels program uses a single-byte instruction discriminator
(``open`` = 1, ``topUp`` = 3), not the 8-byte Anchor discriminator; the
generated builders encode it.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field

from solders.instruction import Instruction  # type: ignore[import-untyped]
from solders.pubkey import Pubkey  # type: ignore[import-untyped]

from pay_kit._paycore.solana import (
    ASSOCIATED_TOKEN_PROGRAM,
    SYSTEM_PROGRAM,
    TOKEN_PROGRAM,
)
from pay_kit.protocols.programs.paymentchannels.instructions.open import Open
from pay_kit.protocols.programs.paymentchannels.instructions.topUp import TopUp
from pay_kit.protocols.programs.paymentchannels.types.distributionEntry import (
    DistributionEntry,
)
from pay_kit.protocols.programs.paymentchannels.types.openArgs import (
    OpenArgs as _OpenArgs,
)
from pay_kit.protocols.programs.paymentchannels.types.topUpArgs import (
    TopUpArgs as _TopUpArgs,
)
from pay_kit.protocols.programs.paymentchannels.types.voucherArgs import VoucherArgs

__all__ = [
    "PAYMENT_CHANNELS_PROGRAM_ID",
    "PROGRAM_ID",
    "Distribution",
    "OpenChannelParams",
    "TopUpParams",
    "build_open_instruction",
    "build_top_up_instruction",
    "find_associated_token_address",
    "find_channel_pda",
    "find_event_authority_pda",
    "voucher_message_bytes",
]

# Canonical payment-channels program id deployed to the network. The IDL
# placeholder ``CQAyft83tN1w2bRofB5PZ79eVDU2xZUVo43LU1qL4zRg`` is NOT the
# production deployment and must not be used for derivation or instruction
# emission.
PAYMENT_CHANNELS_PROGRAM_ID = "GuoKrzaBiZnW5DvJ3yZVE7xHqbcBvaX9SH6P6Cn9gNvc"

#: Parsed production program id used for derivation and instruction emission.
PROGRAM_ID = Pubkey.from_string(PAYMENT_CHANNELS_PROGRAM_ID)

# Channel PDA seed prefix.
_CHANNEL_SEED = b"channel"

# Event-authority PDA seed prefix.
_EVENT_AUTHORITY_SEED = b"event_authority"

# Rent sysvar id.
_RENT_SYSVAR_ID = "SysvarRent111111111111111111111111111111111"


@dataclass(frozen=True)
class Distribution:
    """A single payout recipient and its basis-point share.

    Attributes:
        recipient: The account that receives this share of channel payouts.
        bps: The recipient's share expressed in basis points (1/100th of a
            percent).
    """

    recipient: Pubkey
    bps: int


@dataclass
class OpenChannelParams:
    """Inputs required to build an ``open`` instruction.

    ``token_program`` defaults to the SPL Token program; pass the Token-2022
    program id for Token-2022 mints.

    Attributes:
        payer: The account funding the channel and signing the open.
        payee: The counterparty the channel pays out to.
        mint: The SPL token mint the channel is denominated in.
        authorized_signer: The key authorized to sign vouchers that redeem
            funds from the channel.
        salt: A caller-chosen u64 that disambiguates channels sharing the same
            payer, payee, mint, and signer; part of the channel PDA seeds.
        deposit: The initial token amount deposited into the channel.
        grace_period: Seconds the payee retains to redeem after the channel
            closes before funds are reclaimable by the payer.
        recipients: Optional payout split; each entry's basis points apportion
            the channel's payouts. Empty means a single implicit payee.
        token_program: The token program owning the mint (SPL Token or
            Token-2022).
        program_id: The payment-channels program the instruction targets;
            defaults to the production deployment.
    """

    payer: Pubkey
    payee: Pubkey
    mint: Pubkey
    authorized_signer: Pubkey
    salt: int
    deposit: int
    grace_period: int
    recipients: list[Distribution] = field(default_factory=list)
    token_program: Pubkey = field(default_factory=lambda: Pubkey.from_string(TOKEN_PROGRAM))
    program_id: Pubkey = field(default_factory=lambda: PROGRAM_ID)


@dataclass
class TopUpParams:
    """Inputs required to build a ``topUp`` instruction.

    Attributes:
        payer: The account adding funds to the channel and signing the top-up.
        channel: The channel PDA being funded.
        mint: The SPL token mint the channel is denominated in.
        amount: The token amount to add to the channel's balance.
        token_program: The token program owning the mint (SPL Token or
            Token-2022).
    """

    payer: Pubkey
    channel: Pubkey
    mint: Pubkey
    amount: int
    token_program: Pubkey = field(default_factory=lambda: Pubkey.from_string(TOKEN_PROGRAM))


def voucher_message_bytes(channel_id: Pubkey, cumulative: int, expires_at: int) -> bytes:
    """Return the 48-byte voucher preimage signed by the authorized signer.

    The signer signs this exact byte string to authorize redeeming
    ``cumulative`` lamports from the channel up to ``expires_at``.

    Layout: ``channelId`` (32) || ``cumulativeAmount`` as little-endian u64
    (offset 32) || ``expiresAt`` as little-endian i64 (offset 40). Encoded by
    the generated ``VoucherArgs`` Borsh layout.

    Args:
        channel_id: The channel PDA the voucher authorizes spending from.
        cumulative: The running total amount the voucher authorizes, encoded as
            a little-endian u64.
        expires_at: Unix timestamp after which the voucher is no longer valid,
            encoded as a little-endian i64.

    Raises:
        ValueError: if ``channel_id`` does not encode to exactly 32 bytes.
    """
    channel_bytes = bytes(channel_id)
    if len(channel_bytes) != 32:
        raise ValueError(f"channel id must be exactly 32 bytes, got {len(channel_bytes)}")
    return bytes(
        VoucherArgs.layout.build(
            {
                "channelId": channel_id,
                "cumulativeAmount": cumulative,
                "expiresAt": expires_at,
            }
        )
    )


def find_channel_pda(
    payer: Pubkey,
    payee: Pubkey,
    mint: Pubkey,
    authorized_signer: Pubkey,
    salt: int,
    program_id: Pubkey = PROGRAM_ID,
) -> tuple[Pubkey, int]:
    """Derive the channel PDA, defaulting to the production program id.

    The channel address is fully determined by its payer, payee, mint,
    authorized signer, and salt, so the same inputs always resolve to the same
    channel.

    Seeds: ``["channel", payer, payee, mint, authorizedSigner, salt u64 LE]``.

    Args:
        payer: The account funding the channel.
        payee: The counterparty the channel pays out to.
        mint: The SPL token mint the channel is denominated in.
        authorized_signer: The key authorized to sign vouchers for the channel.
        salt: A caller-chosen u64 disambiguating channels with otherwise
            identical seeds, packed little-endian into the seeds.
        program_id: The payment-channels program to derive against; defaults to
            the production deployment.

    Returns:
        A ``(pubkey, bump)`` pair for the derived channel PDA.
    """
    return Pubkey.find_program_address(
        [
            _CHANNEL_SEED,
            bytes(payer),
            bytes(payee),
            bytes(mint),
            bytes(authorized_signer),
            struct.pack("<Q", salt),
        ],
        program_id,
    )


def find_event_authority_pda(program_id: Pubkey = PROGRAM_ID) -> tuple[Pubkey, int]:
    """Derive the event-authority PDA, defaulting to the production program id.

    The event authority is the program-signed account the program uses to emit
    its CPI event log.

    Seeds: ``["event_authority"]``. Derived here rather than via the generated
    helper because that helper pins the IDL placeholder program id and accepts
    no override.

    Args:
        program_id: The payment-channels program to derive against; defaults to
            the production deployment.

    Returns:
        A ``(pubkey, bump)`` pair for the derived event-authority PDA.
    """
    return Pubkey.find_program_address([_EVENT_AUTHORITY_SEED], program_id)


def find_associated_token_address(
    owner: Pubkey,
    mint: Pubkey,
    token_program: Pubkey,
) -> tuple[Pubkey, int]:
    """Derive the associated token account address for ``(owner, mint, program)``.

    Seeds: ``[owner, token_program, mint]`` under the associated-token program.

    Args:
        owner: The account that owns the token account.
        mint: The SPL token mint the account holds.
        token_program: The token program owning the mint (SPL Token or
            Token-2022).

    Returns:
        A ``(pubkey, bump)`` pair for the derived associated token account.
    """
    return Pubkey.find_program_address(
        [bytes(owner), bytes(token_program), bytes(mint)],
        Pubkey.from_string(ASSOCIATED_TOKEN_PROGRAM),
    )


def build_open_instruction(params: OpenChannelParams) -> Instruction:
    """Build the ``open`` instruction that creates and funds a channel.

    Derives the channel PDA, the payer and channel ATAs, and the event-authority
    PDA, then delegates encoding and the 13 account metas to the generated
    ``Open`` builder against the target program id. The account metas are
    emitted in the fixed order the program expects.

    Args:
        params: The payer, payee, mint, signer, salt, deposit, grace period,
            recipient split, token program, and target program id for the
            channel to open.

    Returns:
        The assembled ``open`` instruction ready to add to a transaction.
    """
    channel, _ = find_channel_pda(
        params.payer,
        params.payee,
        params.mint,
        params.authorized_signer,
        params.salt,
        params.program_id,
    )
    payer_token_account, _ = find_associated_token_address(params.payer, params.mint, params.token_program)
    channel_token_account, _ = find_associated_token_address(channel, params.mint, params.token_program)
    event_authority, _ = find_event_authority_pda(params.program_id)

    args = _OpenArgs(
        salt=params.salt,
        deposit=params.deposit,
        gracePeriod=params.grace_period,
        recipients=[DistributionEntry(recipient=entry.recipient, bps=entry.bps) for entry in params.recipients],
    )
    return Open(
        {"openArgs": args},
        {
            "payer": params.payer,
            "payee": params.payee,
            "mint": params.mint,
            "authorizedSigner": params.authorized_signer,
            "channel": channel,
            "payerTokenAccount": payer_token_account,
            "channelTokenAccount": channel_token_account,
            "tokenProgram": params.token_program,
            "systemProgram": Pubkey.from_string(SYSTEM_PROGRAM),
            "rent": Pubkey.from_string(_RENT_SYSVAR_ID),
            "associatedTokenProgram": Pubkey.from_string(ASSOCIATED_TOKEN_PROGRAM),
            "eventAuthority": event_authority,
            "selfProgram": params.program_id,
        },
        program_id=params.program_id,
    )


def build_top_up_instruction(params: TopUpParams) -> Instruction:
    """Build the ``topUp`` instruction that adds funds to an open channel.

    Derives the payer and channel ATAs, then delegates encoding and the 6
    account metas to the generated ``TopUp`` builder against the production
    program id. The account metas are emitted in the fixed order the program
    expects.

    Args:
        params: The payer, channel, mint, amount, and token program for the
            top-up.

    Returns:
        The assembled ``topUp`` instruction ready to add to a transaction.
    """
    payer_token_account, _ = find_associated_token_address(params.payer, params.mint, params.token_program)
    channel_token_account, _ = find_associated_token_address(params.channel, params.mint, params.token_program)

    return TopUp(
        {"topUpArgs": _TopUpArgs(amount=params.amount)},
        {
            "payer": params.payer,
            "channel": params.channel,
            "payerTokenAccount": payer_token_account,
            "channelTokenAccount": channel_token_account,
            "mint": params.mint,
            "tokenProgram": params.token_program,
        },
        program_id=PROGRAM_ID,
    )
